### ISSUE TITLE
Review of draft-guide-microprofile-openapi for Think

### ISSUE THREAD

**Issue by kwsutter** [OpenLiberty/draft-guide-microprofile-openapi#2]
Attempting https://qa-guides.mybluemix.net/guides/microprofile-openapi.html

The include for org.eclipse.microprofile.openapi.annotations.Operation is missing from InventoryResource.java.  When I performed the `mvn package` after first inserting the various openapi annotations, I got a build error.  I updated my imports and the build worked.

_-> Good catch, fixed._

The order of the resulting openapi output doesn't match between the guide and the actual output when accessing the openapi endpoint.  On my system, the generic inventory/systems is displayed first and the specific inventory/systems/{hostname} is displayed second.  If this ordering is non-deterministic, then maybe a comment in the guide indicating that either one may come first.

_-> Weird, it might be non-deterministic as you say since the order I see them in is reversed. The order may have been switched around in the newer builds, because I remember the order being different a while back. Anyway, I'll ask a few others to try this as well to see what order these endpoints appear in for them._

The description of the OASFactory and OASModelReader sounds like a good exercise.  Maybe this should be a follow-on guide?  If you agree, could you create a separate issue for this request.  Also, as an idea...  Why point them at a TCK example for the OASModelReader?  That seems a bit confusing.   Shouldn't we provide an example OASModelReader that is used with this Inventory service?  Maybe this is also part of this separate guide that I am requesting.  At a minimum, we should highlight that this TCK example is just that -- an example with no relationship to the current guide sample that is being created.

_-> Good point but after getting the SME review we decided to remove this section entirely since using the OASModelReader was an unlikely scenario anyway :)_

The filterAPIResponse is kind of convoluted, isn't it?  Would customers really filter based on the exact description field matching?  Wouldn't it be better to match on the responseCode of 404?  

_-> They probably wouldn't but I was just using "Missing description" as an example._

In the filterOpenAPI() method description, it states "Note that this is the only element which cannot be removed. Note also that this method is always called last for a given filter."  Which element is being referred to here and why can't it be removed?  And, when you reference "this method", are you referring to filterOpenAPI?  

_-> Clarified_

Let's make the reference to "MicroProfile Config" a hotlink to the MP Config github repo or project:  https://github.com/eclipse/microprofile-config

_-> Done, also mentioned that they can read one of our config guides_

Can you explain where the mp.openapi.filter key is used?  Is it a specially defined key used by the OpenAPI implementation?  Or, is it defined in an xml file somewhere?  Or, some java code?

_-> I'm not sure what you mean by this. The key is used in `microprofile-config.properties` and its shown in the guide?_

"repository" is spelled incorrectly:  "To learn more about what elements you can filter, visit the official MicroProfile OpenAPI GitHub repistory:"

_-> fixed_

Isn't kind of strange to point them at actual code for a list of elements that are filterable?  Isn't that part of the javadoc?  The javadoc is hosted on openliberty.io.  Can we point them at that?

_-> Yeah... good point. At the moment, the mp 1.3 api is not yet published on the public site so there is nowhere to point to unfortunately. Lets keep this here for now until the 1.3 api becomes publicly available?_

When talking about disabling the annotation scanning, maybe we could reference the Config feature again?  Something like, "..then annotation scanning can be disabled using the mp.openapi.scan.disable Config property."

_-> Ok good point. I added this in._

For Testing the Service...  It's not clear that the provided tests will work with the annotation scanning disabled with the generated OpenAPI tree (the previous step in the guide).  I stopped the liberty server and ran "mvn install".  Even with the the annotation scanning disabled, I see that two tests were executed and were successful.  So, it's a confusing part of the guide.  I think we should just explain how to execute the full build at the end.  We can still mention that the generated OpenAPI tree is not tested automatically.

_-> It does say "No explicit tests exist to verify the correctness of the generated OpenAPI tree. This can be verified manually by visiting {openapi-url} or {openapi-url}/ui. Though I suppose this can be made clearer by stating it in the first sentence of the section."_

```
mvn liberty:stop-server
mvn install
```

Otherwise, this looks great!  Thanks!

**Comment by Kubik42** [comment]
@kwsutter thanks for your feedback! I've incorporated most of the things here in bd76f334ac8954c4a8eddf28b71cd1b8e775eef8. Do you mind looking over my comments that I left here?

**Comment by Kubik42** [comment]
Do you have any other suggestions @kwsutter ?

**Comment by kwsutter** [comment]
@Kubik42 For some reason, I'm not getting notifications with your updates.  Also, I didn't notice your inline comments...  It would be easier if a new comment was added.  But, now that I have discovered the comments...

The updates in the [bd76f33](https://github.com/OpenLiberty/draft-guide-microprofile-openapi/commit/bd76f334ac8954c4a8eddf28b71cd1b8e775eef8) commit you referenced look good.

So, the whole OASFactory and OASModelReader has been removed?  Must have been an earlier commit.  Sounds like the right plan.

The point about the mp.openapi.filter property key is that we don't talk about how this is used by Config and OpenAPI.  Maybe that's too much and it would get into the Config guide, but Config is powerful and I was hoping for a bit more introduction.  But, it's fine at this point.

How about if we point at the MP javadoc and leave your reference as is for the available filters?  The MP 1.3 javadoc is supposed to go live tomorrow.  Eventually, it will be here...  https://openliberty.io/docs/ref/microprofile/1.3/  Or, you could point them at our generic https://openliberty.io/docs/ reference.

My point about the testing reference at the end is that it leaves the reader hanging...  Are they supposed to try the testing via "mvn install", or will it fail because of the generated OpenAPI tree, or what?  Maybe your re-wording helps in this regard.  I would need to see it in context, I guess.

Thanks,
Kevin

**Comment by Kubik42** [comment]
@kwsutter I'm not sure why you didn't got notified :/ 

Regarding the OASFactory and OASModelReader, we thought it was a very unlikely scenario that users would actually use them to construct the whole OpenAPI document. It might be a topic for a later guide, but for now we decided not to go into detail for these two. OASFactory is still used however in the filter section.

I was thinking of replacing the OASFiler link with the 1.3 API once it goes live. There is actually a comment in the doc already that says to do that :)

As for the tests, I see what you meant. I ended up tweaking the first sentence a bit to say: "A few tests are included for you to test the basic functionality of the inventory service, but they do not test the correctness of the generated OpenAPI document." This should make it clear that the tests exist to only test the service logic and not the OpenAPI doc.
